kubespawner/objects.py

```diff
@@ -46,13 +51,22 @@ def make_pod(
     mem_guarantee=None,
     extra_resource_limits=None,
     extra_resource_guarantees=None,
-    lifecycle_hooks=None,
-    init_containers=None,
+    lifecycle_hooks={},
+    init_containers=[],
     service_account=None,
     extra_container_config=None,
     extra_pod_config=None,
-    extra_containers=None,
-    scheduler_name=None
+    extra_containers=[],
+    scheduler_name=None,
+    tolerations=[],
```

Comment:
Woooosh wow read up on this, yikes!

What about this situation? I'm thinking about the first parameter in the Dict() traitlet. It is another matter since it is not a default parameter, but simply a value acting as a default value. Should be fine to do this initialization right?

```python
    extra_labels = Dict(
        {},
        config=True,
        help="""
        Extra kubernetes labels to set on the spawned single-user pods.

        The keys and values specified here would be set as labels on the spawned single-user
        kubernetes pods. The keys and values must both be strings that match the kubernetes
        label key / value constraints.

        See https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/ for more
        info on what labels are and why you might want to use them!

        `{username}` and `{userid}` are expanded to the escaped, dns-label safe
        username & integer user id respectively, wherever they are used.
        """
    )
```

### Question

If the first thing you do in `jupyterhub_config.py` in the (hub image config that will execute first of all) is to check the type of `c.KubeSpawner.extra_labels`, would you get a traitlets object or a dict object?

If you get the traitlets object, when is `c.KubeSpawner.extra_labels` initialized to `{}`?

Reply:
Okay I've learned more about traitlets now.

c = Config()
Where Config is inheriting from dict. I can do `c.cat_sound = 'meow'` on a config object but not on a dict. I can still use `c.update({'dog_sound': 'barf'})` though. And if i would print `c` it would look like i printed a Dict etc...

So in jupyterhub_config.py we are populating a Config(), but is it blank or is it prepopulated with stuff? When are the KubeSpawner traitlets' default values coming into play? Are they readable from the jupyterhub_config.py already? Hmmm... I'm continuing my investigation.

Reply:
Looking into JupyterHub, I found traitlets with default values being traitlets! This is probably fine to do then. I remember seeing a lot of copy-logic within the traitlet repo as well to strengthen the idea that this will be ok.

From https://github.com/jupyterhub/jupyterhub/blob/master/jupyterhub/app.py#L631-L632
```python
    services = List(Dict(),
        help="""List of service specification dictionaries.
        ...
```

If I'd ensure KubeSpawners traitlets has default values like...
```python
    extra_labels = Dict(
        Dict(),
        config=True,
        help="""
        Extra kubernetes labels to set on the spawned single-user pods.
```

instead of ...

```python
    extra_labels = Dict(
        {},
        config=True,
        help="""
        Extra kubernetes labels to set on the spawned single-user pods.
```

and thereby mimicing the code in JupyterHub, it will be safe and sound, right @minrk ?